iTwin/iTwinUI
# chore: Add Dockerfile with manual chrome install

### What I did

- Add a new `Dockerfile` which is a modified version of the [one from backstopjs](https://github.com/garris/BackstopJS/blob/master/docker/Dockerfile). This Dockerfile has two very important changes:
  1. Uses `bullseye` version of debian through nodejs base image
  2. Manually downloads chromium 100 instead of relying on puppeteer
- Added `build-docker` command to package.json to build platform-specific image from Dockerfile and load into Docker without having publish to Dockerhub.
  - The platform is detected using `os.arch()` in `buildDocker.js`
- Update build workflow to use `docker/build-push-action` for building image from Dockerfile. This takes around 3 mins on CI when uncached and 1 min when cached.
- Update failing test images

### How this helps

Puppeteer is unable to correctly install chromium on M1 Mac (probably because arm version doesn't exist). This change gets around that by explicitly installing chromium.

An additional benefit is that we can now control the chromium version without waiting for backstop/puppeteer to update the deps.

### Additional info

There are 60 updated test images, most of which are probably due to rendering differences between chromium versions, but some of them are because of monospace fonts changing between different linux distros.

In the future, we might want to publish this image to Dockerhub.

### How to test

Run `yarn build-docker`. If this fails, then run `docker buildx create --use` and then `yarn build-docker` again. This should take 3-15 mins depending on your system. This will build and import the image into docker so it needs to be done only once.

After that you can run your tests as usual, e.g. `yarn test` / `yarn test:ci --filter=alert`
